pyproject.toml

```diff
@@ -0,0 +1,47 @@
+[tool.poetry]
+name = "firefox-translations-training"
+version = "0.1.0"
+description = ""
+authors = []
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.10"
+
+[tool.poetry.group.black.dependencies]
+black = "^23.7.0"
+
+
+[tool.poetry.group.lint.dependencies]
+ruff = "^0.0.287"
+
+[tool.black]
+extend-exclude= "/3rd_party"
+
+# Match the rules from mozilla-central.
+# https://searchfox.org/mozilla-central/source/pyproject.toml
+[tool.ruff]
```

Comment:
How do we integrate this all with Pycharm/Vscode? Can we move those rules to a hidden file in a way it's done with [flake8](https://github.com/psf/black/blob/main/.flake8)? The workflow with setting up a poetry env and running the makefile every time before push looks a bit manual to me. Especially since I often use a different conda env for debugging things locally. Ideally, the IDE should use the linter and the formatter too. Then I would be able to autofix issues and to reformat manually with a hot key.

Reply:
> How do we integrate this all with Pycharm/Vscode?

In VS Code you can set up: `.vscode/settings.json` with the following rule and everything will auto-format. You can also do it via shortcut with a somewhat different commit.

```
{
  "[python]": {
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
      "source.fixAll": true
    }
  },
  "python.formatting.provider": "black"
}
```

> ? Can we move those rules to a hidden file in a way it's done with [flake8](https://github.com/psf/black/blob/main/.flake8)?

Python in general is converging on using a `pyproject.toml` at the root folder of your project to define the environment and dependencies for your project: https://peps.python.org/pep-0518/

In fact this is the way [mozilla-central is configuring things](https://searchfox.org/mozilla-central/source/pyproject.toml).

> The workflow with setting up a poetry env and running the makefile every time before push looks a bit manual to me.

I really prefer the automatic editor saving as the mechanism. This is what the recommendation is for mozilla-central by those that manage linting [citation needed]. There is the pre-commit hook mechanism, but that feels a bit invasive to me. In mozilla-central there is a `mach lint --fix -wo` command which can be run before pushing. I could add a new makefile command called `make fix-all` to combine it all into one command.

> Especially since I often use a different conda env for debugging things locally. Ideally, the IDE should use the linter and the formatter too

I think the poetry decision is up for debate. I ended up going with it since it's an easy way to manage dependencies, and generate a lockfile. It handles the virtualenv isolation for running tasks like this. It does not handle python versioning.

Our solution should:

* Generate a lockfile for consistent CI runs.
* Provide local isolation for running an environment.

To me it feels safer than some other solutions, as it doesn't require messing with your local environment as much. I get a bit nervous as my python configuration breaks across projects, and poetry feels fairly self-contained, compared to others I have used in the past.

Reply:
The integration with vscode looks easier than with pycharm: https://black.readthedocs.io/en/stable/integrations/editors.html.

Anyway, there should be a way to autoformat with the IDE. I'm okay with using poetry fir CI or having an ability to run it locally with make but the rules we override should be stored where the IDE plugins can find them, otherwise, you'll need to reformat two times, one with the IDE and the final one to follow the actual rules. Will the IDE plugins respect the rules from the pyproject.toml? Or do we need to extract them to another place?

Reply:
I looked at some popular python open-source projects that use Ruff and they also specify rules in the pyproject.toml. (https://github.com/huggingface/transformers, https://github.com/pandas-dev/pandas, https://github.com/tiangolo/fastapi/blob/master/pyproject.toml). It does look like a standard now. Let's keep it that way then and figure out how to make IDEs respect that.

Reply:
> Will the IDE plugins respect the rules from the pyproject.toml?

I can say yes for VS Code, but I don't know for other editors. I will say that the `pyproject.toml` is what the python ecosystem is standardizing on, so editors _should_ be able to read and understand it, especially if they have an active plugin system.